### PR TITLE
refactor: agenttools facade uses model config service

### PR DESCRIPTION
**Pull request by barrettj12**
The agenttools facade now uses a model config service to access model config, and a model agent service to access the current agent version (as introduced in #17598). Unfortunately, we can't entirely remove state from this facade yet, as we still need it to update the model.

We introduce a new interfaces `ModelConfigService` and `ModelAgentService`, stored in `service.go`, and replaced with gomock-generated mocks in testing.

BREAKING CHANGE: agenttools now checks all preferred streams in preference order, starting with the user-specified stream in `agent-stream`. Previously, it always checked the "released" stream first, even if a custom `agent-stream` was set, and it would only check the user-defined stream if nothing was found in "released".

Additional changes:
- Add a new env variable to the Makefile (`JUJU_PUBLISH_STREAM`) that allows `make simplestreams` to publish to a specified stream.
- General modernising and code improvement on the agenttools facade, such as moving internal functions to methods on the API struct.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

This facade reads the following keys from model config: `agent-version`, `agent-stream`, `development`. `agent-version` is immutable, but the other two can be changed and observed to have an effect.

By default, the toolsversionchecker worker checks for updated tools every six hours. To QA this, we need to set this value to something much smaller, e.g. 5 seconds:
https://github.com/juju/juju/blob/65266db76c2f84e4a545a574d4db82d3388d89eb/internal/worker/toolsversionchecker/manifold.go#L47-L49

Publish simplestreams data into "proposed" and "devel" streams:
```bash
# set version to 4.0-beta5
$ JUJU_PUBLISH_STREAM=proposed make simplestreams
# set version to 4.0-beta5
$ JUJU_PUBLISH_STREAM=devel make simplestreams
# set version back to 4.0-beta4
$ make go-build
```

Start the simplestreams server:
```
$ cd _build/simplestreams/tools
$ python3 -m http.server
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
```

Use `hostname -I | cut -d' ' -f1` to get your computer's IP address, we will feed this to Juju config.

Bootstrap Juju with the given simplestreams server, and `agent-stream` initially set to "proposed":
```bash
juju bootstrap lxd c \
  --config agent-metadata-url=http://192.168.1.25:8000/ \
  --config agent-stream=proposed
```

Switch to the `controller` model and run `juju status`. We should see that an upgrade to 4.0-beta5 is available.
```
$ juju status
Model       Controller  Cloud/Region         Version      Timestamp       Notes
controller  c           localhost/localhost  4.0-beta4.1  17:02:06+12:00  upgrade available: 4.0-beta5
```

Now change the agent stream to `devel`, wait for the toolsversionchecker worker to update, and check that an upgrade to 4.0-beta6 is available.
```
$ juju model-config agent-stream=devel
$ juju status
Model       Controller  Cloud/Region         Version      Timestamp       Notes
controller  c           localhost/localhost  4.0-beta4.1  17:02:06+12:00  upgrade available: 4.0-beta6
```

You can do the same with the `development` model config key and the `devel` stream, however observe that this key only has an effect for non-development versions of Juju, so you would have to try it with versions e.g. 4.0.0 and 4.0.1.

## Links

**Jira card:** JUJU-6104

### CODE REVIEW

File: apiserver/facades/controller/agenttools/agenttools.go [hunk 1]
```diff
@@ -94,21 +98,21 @@ func checkToolsAvailability(ctx context.Context, newEnviron newEnvironFunc, mode
...
 	return env.UpdateLatestToolsVersion(ver)
 }
 
-func updateToolsAvailability(ctx context.Context, modelGetter ModelGetter, newEnviron newEnvironFunc, finder toolsFinder, update envVersionUpdater, logger corelogger.Logger) error {
-	model, err := modelGetter.Model()
-	if err != nil {
-		return errors.Annotate(err, "cannot get model")
-	}
-	cfg, err := modelConfig(model)
+func updateToolsAvailability(
```
Comment (tlm) [inline apiserver/facades/controller/agenttools/agenttools.go:106]: We are going to need to make some changes here.

Specifically using model config to get the models agent version is something we are actively stamping out.

At the moment we have been replicating the agent version set on the agent version table in the controller database into the read only model information. This was a bad design decision that we have to undo now.

What I propose as part of this PR is that we introduce a new service for interacting with a models agent information. We can nut out the design together and quickly get it implemented.

Comment (tlm) [reply apiserver/facades/controller/agenttools/agenttools.go:106]: I am also thinking that we can avoid passing all the services into these functions and just load them up with the information they need.

**Review by tlm — COMMENTED** [review]
Thanks @barrettj12, this is good but think we need to nail down the agent version stuff a little more.

File: apiserver/facades/controller/agenttools/agenttools.go (hunk 1 above)
Comment (barrettj12) [reply apiserver/facades/controller/agenttools/agenttools.go:106]: Honestly this whole facade probably needs to be redone, it's using some bad design patterns, but I don't know if we should do that in this PR.

Comment (barrettj12) [reply apiserver/facades/controller/agenttools/agenttools.go:106]: P.S. are you talking about pulling some of this facade logic into the new service?

**Comment by barrettj12** [comment]
#17598 needs to merge before this

File: apiserver/facades/controller/agenttools/agenttools.go [hunk 2]
```diff
@@ -80,10 +86,18 @@ func checkToolsAvailability(ctx context.Context, newEnviron newEnvironFunc, mode
...
+		var modelCfg *config.Config
+		modelCfg, err = api.modelConfigService.ModelConfig(ctx)
+		if err != nil {
+			return version.Zero, errors.Annotate(err, "cannot get config")
+		}
+
+		preferredStream := tools.PreferredStreams(&currentVersion, modelCfg.Development(), modelCfg.AgentStream())[0]
+		if preferredStream != tools.ReleasedStream {
+			vers, err = api.findTools(ctx, ss, env, currentVersion.Major, currentVersion.Minor, []string{preferredStream}, coretools.Filter{})
+		}
```
Comment (barrettj12) [inline apiserver/facades/controller/agenttools/agenttools.go:100]: Honestly I don't think this logic is correct, I think we should be getting the list of `preferredStreams` first from `tools.PreferredStreams`, and then feeding the whole list to `api.findTools`.

@wallyworld thoughts? Maybe I'm missing something here as I don't know why we did it this way in the first place.

File: apiserver/facades/controller/agenttools/service.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+// ModelConfigService provides access to the model configuration.
+type ModelConfigService interface {
+	// ModelConfig returns the current config for the model.
+	ModelConfig(context.Context) (*config.Config, error)
+}
+
+type ModelAgentService interface {
+	GetModelAgentVersion(ctx context.Context) (version.Number, error)
```
Comment (tlm) [inline apiserver/facades/controller/agenttools/service.go:21]: Comments please.

**Review by tlm — APPROVED** [review]
Few small nits

File: apiserver/facades/controller/agenttools/agenttools.go [hunk 4]
```diff
@@ -63,13 +66,16 @@ type newEnvironFunc func() (environs.Environ, error)
...
 
-func checkToolsAvailability(ctx context.Context, newEnviron newEnvironFunc, modelCfg *config.Config, finder toolsFinder) (version.Number, error) {
-	currentVersion, ok := modelCfg.AgentVersion()
-	if !ok || currentVersion == version.Zero {
+func (api *AgentToolsAPI) checkToolsAvailability(ctx context.Context) (version.Number, error) {
+	currentVersion, err := api.modelAgentService.GetModelAgentVersion(ctx)
+	if err != nil {
+		return version.Zero, errors.Annotate(err, "getting agent version from service")
+	}
+	if currentVersion == version.Zero {
```
Comment (wallyworld) [inline apiserver/facades/controller/agenttools/agenttools.go:74]: I wonder why we checked for version.Zero. The model agent version is either known or it's an error. I think the new agent service should never return a Zero version without an error, in which case we don't need this check.

File: apiserver/facades/controller/agenttools/agenttools.go [hunk 5]
```diff
@@ -63,13 +66,16 @@ type newEnvironFunc func() (environs.Environ, error)
...
 		return version.Zero, nil
 	}
 
-	env, err := newEnviron()
+	env, err := api.newEnviron()
 	if err != nil {
 		return version.Zero, errors.Annotatef(err, "cannot make model")
```
Comment (wallyworld) [inline apiserver/facades/controller/agenttools/agenttools.go:80]: strictly in this case it's an Environ, which we surface as "cloud provider"
so maybe "cannot make cloud provider" or something

File: apiserver/facades/controller/agenttools/agenttools.go (hunk 2 above)
Comment (wallyworld) [reply apiserver/facades/controller/agenttools/agenttools.go:100]: I think the original logic is clumsy and inefficient but the intent seems in line with what we want, ie search for agents in order of devel, proposed, released; first match wins. FindTools handles this - I suspect the code here was done earlier and was never refactored. As you say, we just pass into FindTools the list of streams to search, preferred ones first in the list; one call to FindTools.

**Review by wallyworld — COMMENTED** [review]
I think we can make one call to FindTools with a  list of streams

File: apiserver/facades/controller/agenttools/agenttools.go (hunk 2 above)
Comment (barrettj12) [reply apiserver/facades/controller/agenttools/agenttools.go:100]: Great, I'll make this change. Technically it's a behaviour change, but I think it's bringing the behaviour more in line with what a user would expect, plus I doubt anyone's depending on the old behaviour

**Pull request merged**